pandas/core/reshape/merge.py

```diff
@@ -2072,9 +2072,13 @@ def _items_overlap_with_suffix(left: Index, right: Index, suffixes: Tuple[str, s
     If corresponding suffix is empty, the entry is simply converted to string.
 
     """
-    if not isinstance(suffixes, tuple):
-        raise TypeError(
-            f"suffixes should be tuple of (str, str). But got {type(suffixes).__name__}"
+    if isinstance(suffixes, set):
+        warnings.warn(
```

Comment:
why not just raise? also what about a dict or non-list-like for example?

do we check is_list_like anywhere?

Reply:
I think we'll need a warning according to our backwards compatibility poilcy.

Reply:
It's easy to do it with a warning, so I would also just do that.

@TomAugspurger if we don't want to have it set-specific, could also do `if not is_list_like(suffixes, allow_sets=False)`, although in practice that might give the same.

Reply:
I agree that should use what @jorisvandenbossche suggests here , e.g. dict would pass this test but is not acceptable

Reply:
dicts are ordered though, so they shouldn't be a problem. I thought the original issue was non-deterministic outputs because sets aren't ordered?

Reply:
actually sets are not ordered, but I think we need an actual sequence here (sure .keys() satisfies), but i'd rather be explicit. we do not treat dicts as list-like anywhere else.